src/modules/ssh_apt/mappers/index.ts

```diff
@@ -0,0 +1,219 @@
+      } = {};
+      for (const e of es) {
+        const packageGroup = packageGroupsByServer[e.server] || {
+          toRestore: [],
+          toInstall: [],
+          toUninstall: [],
+        };
+        const id = this.generateId({
+          server: e.server,
+          package: e.package,
+          version: e.version,
+          architecture: e.architecture,
+        });
+        const cloudE = ctx.memo.cloud.Package[id] as Package;
+        if (e.description !== cloudE.description || e.upgradable !== cloudE.upgradable) {
+          cloudE.id = e.id;
+          packageGroup.toRestore.push(cloudE);
+        } else if (e.installed) {
+          // We know this is the only other column, so it must have changed
+          packageGroup.toInstall.push(e);
+        } else {
+          packageGroup.toUninstall.push(e);
+        }
+      }
+      const clientsAndServers = await Promise.all(
+        Object.keys(packageGroupsByServer).map(async (s: string) => ({
+          server: s,
+          client: await ctx.getSshClient(s),
+        })),
+      );
+      await Promise.all(
+        clientsAndServers.map(async clientAndServer => {
+          const { client, server } = clientAndServer;
+          const packageGroup = packageGroupsByServer[server];
+          if (packageGroup.toRestore.length) {
+            await this.db.update(packageGroup.toRestore, ctx);
+          }
+          if (packageGroup.toUninstall.length) {
+            const removeCommand = `sudo apt remove ${packageGroup.toUninstall.map(e => e.package).join(' ')}`;
+            try {
+              await client.exec(removeCommand);
+            } catch (e) {
+              // stderr causes a throw in this API, so it's not important; we will also explicitly
+              // sync the db with the cloud state once done, anyways
+            }
+          }
+          if (packageGroup.toInstall.length) {
+            const installCommand = `sudo apt install ${packageGroup.toInstall
+              .map(e => `${e.package}=${e.version}`)
+              .join(' ')}`;
+            try {
+              await client.exec(installCommand);
+            } catch (e) {
+              // stderr causes a throw in this API, so it's not important; we will also explicitly
+              // sync the db with the cloud state once done, anyways
+            }
+          }
+        }),
+      );
+      // Flush the cache to be sure we're re-reading from the servers
+      delete ctx.memo.cloud.Package;
+      await this.cloud.read(ctx);
+      // For this pass, only worry about the packages that were marked to be updated and replace
+      // only those in the database
+      const newPackages = [];
+      for (const e of es) {
+        const id = this.generateId({
+          server: e.server,
+          package: e.package,
+          version: e.version,
+          architecture: e.architecture,
+        });
+        const newPackage = (await this.cloud.read(ctx, id)) as Package;
```

Comment:
nit: can use `cloud.read(ctx)` without the `id` field and do the filtering here instead?

Reply:
This keeps it DRY instead? (And the `memo` cache should be hit so it doesn't actually make another call to the server.)